src/proposal.mligo

```diff
@@ -277,84 +282,93 @@ let do_total_vote_meet_quorum_threshold (proposal, store: proposal * storage): b
+  let ops = List.fold cons ops new_ops in
+  (ops, store)
+
+
+(*
+ * Flush a proposal, starting from the first. Remove it when the flush is done, and move to the next proposal.
+ *)
+let rec flush_each (n, config, store, ops : int * config * storage * operation list): (operation list * storage * int) =
+  let (plist_head, plist_new) = plist_pop store.ongoing_proposals_dlist in
+  // Ensure plist is not empty.
+  match plist_head with
+    | Some first_key ->
+        let proposal = check_if_proposal_exist (first_key, store) in
+        // Ensure the proposal is not expired.
+        let _ = if is_proposal_age (proposal, config.proposal_expired_level)
+            then (failwith expired_proposal : unit)
+            else unit in
 
-  if is_proposal_age (proposal, config.proposal_expired_level)
-  then (failwith expired_proposal : (operation list * storage * counter))
-  else if is_proposal_age (proposal, config.proposal_flush_level)
-       && counter.current < counter.total // not finished
-  then
-    let counter = { counter with current = counter.current + 1n } in
-    let cond =    do_total_vote_meet_quorum_threshold(proposal, store)
-              && proposal.upvotes > proposal.downvotes
-    in
-    let store = unstake_proposer_token
-          (config.rejected_proposal_slash_value, cond, proposal, config.period, config.fixed_proposal_fee_in_token, store) in
-    let (new_ops, store) =
-      if cond
-      then
-        let dl_out = config.decision_lambda { proposal = proposal; extras = store.extra } in
-        let guardian = match dl_out.guardian with
-          | Some g -> g
-          | None -> store.guardian
-        in (dl_out.operations,
-              { store with extra = dl_out.extras
-              ; guardian = guardian
-              })
-      else (nil_op, store)
-    in
-    let cons = fun (l, e : operation list * operation) -> e :: l in
-    let ops = List.fold cons ops new_ops in
-    let store = remove_from_proposal_sort_by_level (start_level, proposal_key, store) in
-    (ops, store, counter)
-  else (ops, store, counter)
+        // Ensure counter is not reached yet and the the proposal is flushable.
+        if (n > 0 && is_proposal_age (proposal, config.proposal_flush_level)) then
+          // Unstake tokens, and execute proposal's decision_lambda.
+          let (ops, store) = handle_proposal_is_over (config, proposal, store, ops) in
+          // Update the stored plist.
+          let store = { store with ongoing_proposals_dlist = plist_new } in
+          // Continue to the next proposal
+          flush_each (n - 1, config, store, ops)
+        else
+          (ops, store, n)
+    | None ->
+        (ops, store, n)
 
-// Flush all proposals that passed their voting stage.
+(*
+ * Flush all proposals that passed their voting stage.
+ *)
 let flush(n, config, store : nat * config * storage): return =
   if n = 0n
   then (failwith empty_flush : return)
   else
```

Comment:
Couldn't this be removed now?

IIUC `flush_each` will return right away if `n` is `0` and other check after that will fail with `empyt_flush` just the same, right?

Reply:
Make sense, I've updated this.